File: JasSwift/Sources/Interpreter/StateStore.swift
/// Reactive state store for the workspace interpreter.
///
/// Manages global state, panel-scoped state, and dialog-scoped state.
/// Pure Swift, no SwiftUI dependency. Port of workspace_interpreter/state_store.py.

import Foundation

/// Manages global state, panel-scoped state, dialog-scoped state, and
/// builds evaluation contexts for the expression evaluator.
public class StateStore {

    // MARK: - Global state

    private var state: [String: Any]
    private var panels: [String: [String: Any]] = [:]
    private var activePanel: String?

    // MARK: - Tool state
    //
    // Parallels `panels` but keyed by tool id. YAML tool handlers
    // read/write via `$tool.<id>.<key>`; the `set` effect routes
    // those targets here via [`setTool`]. Populated when YamlTool
    // (Phase 5) constructs with a tool spec and seeds its defaults.

    private var tools: [String: [String: Any]] = [:]

    // MARK: - Dialog state

    private var dialog: [String: Any] = [:]
    private var dialogId: String?
    private var dialogParams: [String: Any]?
    private var dialogProps: [String: [String: Any]] = [:]  // {key: {"get": expr, "set": expr}}
    /// Captured original values of state keys named in the open dialog's
    /// preview_targets. Restored on closeDialog (via the close_dialog
    /// effect) unless first cleared by the clear_dialog_snapshot effect
    /// (used by OK actions).
    private var dialogSnapshot: [String: Any]?
    /// Action name fired by the post-run hook in runEffects when the
    /// dialog mutates. Set by openDialog from the dialog spec's
    /// on_change field. Cleared on closeDialog.
    private var dialogOnChange: String?
    /// Set by every setDialog write; consumed by the post-run hook
    /// to decide whether to fire the dialog's on_change action.
    private var dialogDirty: Bool = false
    /// Re-entrancy guard: true while the post-run hook is dispatching
    /// the on_change action. Prevents the action's own set effects
    /// from re-triggering the hook.
    private var dialogFiringOnChange: Bool = false

    // MARK: - Data namespace
    //
    // Workspace-loaded reference data (swatch_libraries,
    // brush_libraries, etc.). Mirrors the JS-side `data` namespace
    // from store.mjs and the Rust StateStore.data field. Mutated by
    // the brush.* effect handlers and read by the canvas brush
    // registry sync.

    private var data: [String: Any] = [:]

    public func setData(_ data: [String: Any]) {
        self.data = data
    }

    public func dataAll() -> [String: Any] {
        data
    }

    /// Read a dotted path inside `data`. Path may include the
    /// "data." prefix or omit it. Returns nil for any missing
    /// intermediate.
    public func getDataPath(_ rawPath: String) -> Any? {
        let path = rawPath.hasPrefix("data.") ? String(rawPath.dropFirst(5)) : rawPath
        if path.isEmpty { return data }
        var cur: Any = data
        for seg in path.split(separator: ".") {
            guard let dict = cur as? [String: Any], let next = dict[String(seg)] else {
                return nil
            }
            cur = next
        }
        return cur
    }

    /// Write a value at a dotted path inside `data`. Intermediate
    /// dicts are created on demand.
    public func setDataPath(_ rawPath: String, _ value: Any?) {
        let path = rawPath.hasPrefix("data.") ? String(rawPath.dropFirst(5)) : rawPath
        if path.isEmpty {
            if let map = value as? [String: Any] { data = map }
            return
        }
        let segs = path.split(separator: ".").map(String.init)
        // Walk-and-build. Build a new dictionary tree by reading from
        // root, splicing in the change at the leaf, and writing back.
        data = setPathInDict(data, segs: segs, value: value)
    }

    private func setPathInDict(_ dict: [String: Any], segs: [String], value: Any?) -> [String: Any] {
        var result = dict
        guard let head = segs.first else { return result }
        if segs.count == 1 {
            if let v = value {
                result[head] = v
            } else {
                result.removeValue(forKey: head)
            }
            return result
        }
        let inner = (result[head] as? [String: Any]) ?? [:]
        result[head] = setPathInDict(inner, segs: Array(segs.dropFirst()), value: value)
        return result
    }

    // MARK: - Init

    public init(defaults: [String: Any]? = nil) {
        self.state = defaults ?? [:]
    }

    // MARK: - Global state accessors

    public func get(_ key: String) -> Any? {
        state[key]
    }

    public func set(_ key: String, _ value: Any?) {
        state[key] = value
    }

    public func getAll() -> [String: Any] {
        state
    }

    // MARK: - Panel state

    public func initPanel(_ panelId: String, defaults: [String: Any]) {
        panels[panelId] = defaults
    }

    public func hasPanel(_ panelId: String) -> Bool {
        panels[panelId] != nil
    }

    public func getPanel(_ panelId: String, _ key: String) -> Any? {
        panels[panelId]?[key]
    }

    public func setPanel(_ panelId: String, _ key: String, _ value: Any?) {
        panels[panelId]?[key] = value
    }

    public func getPanelState(_ panelId: String) -> [String: Any] {
        panels[panelId] ?? [:]
    }

    public func setActivePanel(_ panelId: String?) {
        activePanel = panelId
    }

    public func getActivePanelId() -> String? {
        activePanel
    }

    public func getActivePanelState() -> [String: Any] {
        guard let pid = activePanel else { return [:] }
        return panels[pid] ?? [:]
    }

    public func destroyPanel(_ panelId: String) {
        panels.removeValue(forKey: panelId)
        if activePanel == panelId {
            activePanel = nil
        }
    }

    // MARK: - Tool state

    public func initTool(_ toolId: String, defaults: [String: Any]) {
        tools[toolId] = defaults
    }

    public func hasTool(_ toolId: String) -> Bool {
        tools[toolId] != nil
    }

    public func getTool(_ toolId: String, _ key: String) -> Any? {
        tools[toolId]?[key]
    }

    public func setTool(_ toolId: String, _ key: String, _ value: Any?) {
        // Create the tool namespace on first write — less friction for
        // callers that haven't explicitly run initTool. Mirrors the Rust
        // set_tool behavior.
        if tools[toolId] == nil {
            tools[toolId] = [:]
        }
        tools[toolId]?[key] = value
    }

    public func getToolState(_ toolId: String) -> [String: Any] {
        tools[toolId] ?? [:]
    }

    public func destroyTool(_ toolId: String) {
        tools.removeValue(forKey: toolId)
    }

    /// Return the whole tool scope (for tests / inspection).
    public func getToolScopes() -> [String: [String: Any]] {
        tools
    }

    // MARK: - Dialog state

    public func initDialog(_ dialogId: String, defaults: [String: Any],
                           params: [String: Any]? = nil,
                           props: [String: [String: Any]]? = nil) {
        self.dialogId = dialogId
        self.dialog = defaults
        self.dialogParams = params
        self.dialogProps = props ?? [:]
    }

    /// Get a dialog value, evaluating the getter if present.
    ///
    /// Dialog getters may reference cross-scope bindings (panel /
    /// state / active_document / param). Pass them via ``outer``.
    /// Artboard Options reference-point transforms require this.
    public func getDialogWithOuter(_ key: String, outer: [String: Any]) -> Any? {
        guard dialogId != nil else { return nil }
        if let prop = dialogProps[key], let getExpr = prop["get"] as? String {
            var local = outer
            for (k, v) in dialog { local[k] = v }
            let result = evaluate(getExpr, context: local)
            return result.toAny()
        }
        return dialog[key]
    }

    public func getDialog(_ key: String) -> Any? {
        getDialogWithOuter(key, outer: [:])
    }

    /// Set a dialog value, running the setter lambda if present.
    /// ``outer`` gives the setter access to cross-scope bindings.
    public func setDialogWithOuter(_ key: String, _ value: Any?, outer: [String: Any]) {
        guard dialogId != nil else { return }
        dialogDirty = true
        if let prop = dialogProps[key] {
            if let setExpr = prop["set"] as? String {
                // Parse the setter as a lambda and apply with the value
                var local = outer
                for (k, v) in dialog { local[k] = v }
                let storeCb: (String, Value) -> Void = { [weak self] target, val in
                    self?.dialog[target] = val.toAny()
                }
                local["__store_cb__"] = storeCb
                let setterVal = evaluate(setExpr, context: local)
                if case .closure(let params, let body, let captured) = setterVal {
                    if params.count == 1 {
                        var callCtx = captured
                        for (k, v) in local { callCtx[k] = v }
                        callCtx[params[0]] = value
                        let _ = evalNode(body, callCtx)
                    }
                }
                return
            }
            if prop["get"] != nil {
                return  // read-only prop — ignore writes
            }
        }
        dialog[key] = value
    }

    public func setDialog(_ key: String, _ value: Any?) {
        setDialogWithOuter(key, value, outer: [:])
    }

    public func getDialogState() -> [String: Any] {
        dialog
    }

    public func getDialogId() -> String? {
        dialogId
    }

    public func getDialogParams() -> [String: Any]? {
        dialogParams
    }

    public func closeDialog() {
        dialogId = nil
        dialog = [:]
        dialogParams = nil
        dialogProps = [:]
        dialogOnChange = nil
        dialogDirty = false
        // dialogFiringOnChange is intentionally NOT cleared here —
        // closeDialog can be invoked from inside an on_change-fired
        // chain, and the guard must remain set until runEffects
        // unwinds.
    }

    public func setDialogOnChange(_ action: String?) {
        dialogOnChange = action
    }

    public func getDialogOnChange() -> String? { dialogOnChange }

    /// Take the dirty flag, leaving it false. Used by the post-run
    /// hook in runEffects to decide whether to fire on_change.
    public func takeDialogDirty() -> Bool {
        let was = dialogDirty
        dialogDirty = false
        return was
    }

    public func isFiringOnChange() -> Bool { dialogFiringOnChange }
    public func setFiringOnChange(_ firing: Bool) { dialogFiringOnChange = firing }

    /// Capture the current value of every state key referenced by a
    /// dialog's preview_targets. Phase 0 supports only top-level state
    /// keys (no dots in the path); deep paths are silently skipped and
    /// will land alongside their first real consumer in Phase 8/9.
    /// `targets` maps `dialog_state_key` → `state_key`.
    public func captureDialogSnapshot(_ targets: [String: String]) {
        var snap: [String: Any] = [:]
        for stateKey in targets.values {
            if !stateKey.contains(".") {
                snap[stateKey] = state[stateKey] as Any
            }
        }
        dialogSnapshot = snap
    }

    public func getDialogSnapshot() -> [String: Any]? {
        dialogSnapshot
    }

    public func clearDialogSnapshot() {
        dialogSnapshot = nil
    }

    public func hasDialogSnapshot() -> Bool {
        dialogSnapshot != nil
    }

    // MARK: - List operations

    public func listPush(_ panelId: String, _ key: String, _ value: Any,
                         unique: Bool = false, maxLength: Int? = nil) {
        guard panels[panelId] != nil else { return }
        var lst = (panels[panelId]?[key] as? [Any]) ?? []
        if unique {
            lst.removeAll { item in
                "\(item)" == "\(value)"
            }
        }
        lst.insert(value, at: 0)
        if let max = maxLength, lst.count > max {
            lst = Array(lst.prefix(max))
        }
        panels[panelId]?[key] = lst
    }

    // MARK: - Context for expression evaluation

    /// Build an evaluation context dict for the expression evaluator.
    /// Returns ["state": [...], "panel": [...], ...].
    /// When a dialog is open, also includes "dialog" and "param" keys.
    public func evalContext(extra: [String: Any]? = nil) -> [String: Any] {
        var ctx: [String: Any] = ["state": state]

        if let pid = activePanel, let scope = panels[pid] {
            ctx["panel"] = scope
        } else {
            ctx["panel"] = [String: Any]()
        }

        // Tool scope — one nested dict per registered tool. Expressions
        // read as `tool.<id>.<key>`.
        ctx["tool"] = tools

        if dialogId != nil {
            ctx["dialog"] = dialog
            if let params = dialogParams {
                ctx["param"] = params
            }
        }

        if let extra = extra {
            for (k, v) in extra {
                ctx[k] = v
            }
        }
        return ctx
    }
}

File: jas_dioxus/src/interpreter/state_store.rs
//! Reactive state store — Rust port of workspace_interpreter/state_store.py.
//!
//! Manages global state, panel-scoped state, and change notifications.
//! Uses callbacks (closures) for reactivity — the Dioxus layer connects
//! these to signal updates.

use std::collections::HashMap;
use serde_json;
use super::expr_types::Value;

/// Reactive state store for global, panel-scoped, tool-scoped, and
/// dialog-scoped state. Mirrors `jas_flask/static/js/engine/store.mjs`
/// except for reactivity (handled by Dioxus signals at the UI layer).
pub struct StateStore {
    state: HashMap<String, serde_json::Value>,
    panels: HashMap<String, HashMap<String, serde_json::Value>>,
    active_panel: Option<String>,
    /// Per-tool local state. YAML handlers read/write via `$tool.<id>.<key>`
    /// paths; the `set` effect routes prefixed targets here via
    /// [`StateStore::set_tool`].
    tools: HashMap<String, HashMap<String, serde_json::Value>>,
    dialog: Option<HashMap<String, serde_json::Value>>,
    dialog_id: Option<String>,
    dialog_params: Option<HashMap<String, serde_json::Value>>,
    /// Captured original values of state keys named in the open dialog's
    /// `preview_targets`. Restored on close_dialog unless first cleared
    /// by the `clear_dialog_snapshot` effect (used by OK actions).
    dialog_snapshot: Option<HashMap<String, serde_json::Value>>,
    /// Action name fired by the post-run hook in `run_effects` when
    /// the dialog mutates. Set by `open_dialog` from the dialog
    /// spec's `on_change` field. Cleared on `close_dialog`.
    dialog_on_change: Option<String>,
    /// Set by every `set_dialog` write; consumed by the post-run hook
    /// to decide whether to fire the dialog's `on_change` action.
    dialog_dirty: bool,
    /// Re-entrancy guard: true while the post-run hook is dispatching
    /// the on_change action. Prevents the action's own `set` effects
    /// (or any nested run_effects calls) from re-triggering the hook.
    dialog_firing_on_change: bool,
    /// Workspace-loaded reference data (swatch_libraries,
    /// brush_libraries, etc.). Mirrors the JS-side `data` namespace
    /// from jas_flask/static/js/engine/store.mjs. Mutated by the
    /// data.* and brush.* effect handlers.
    data: serde_json::Value,
}

impl StateStore {
    pub fn new() -> Self {
        StateStore {
            state: HashMap::new(),
            panels: HashMap::new(),
            active_panel: None,
            tools: HashMap::new(),
            dialog: None,
            dialog_id: None,
            dialog_params: None,
            dialog_snapshot: None,
            dialog_on_change: None,
            dialog_dirty: false,
            dialog_firing_on_change: false,
            data: serde_json::Value::Object(serde_json::Map::new()),
        }
    }

    /// Initialise the data namespace from a JSON object. Replaces
    /// any existing data. App startup typically calls this with the
    /// loaded workspace so data.brush_libraries / data.swatch_libraries
    /// resolve in YAML expressions.
    pub fn set_data(&mut self, data: serde_json::Value) {
        self.data = data;
    }

    pub fn data(&self) -> &serde_json::Value {
        &self.data
    }

    /// Read a dotted path inside `data`. Path may include the
    /// `data.` prefix or omit it; both walk the same tree. Returns
    /// `Value::Null` for any missing intermediate.
    pub fn get_data_path(&self, raw_path: &str) -> serde_json::Value {
        let path = raw_path.strip_prefix("data.").unwrap_or(raw_path);
        if path.is_empty() {
            return self.data.clone();
        }
        let mut cur = &self.data;
        for seg in path.split('.') {
            match cur.get(seg) {
                Some(v) => cur = v,
                None => return serde_json::Value::Null,
            }
        }
        cur.clone()
    }

    /// Write a value at a dotted path inside `data`. Intermediate
    /// objects are created on demand. Numeric segments index into
    /// arrays when the parent is already an array.
    pub fn set_data_path(&mut self, raw_path: &str, value: serde_json::Value) {
        let path = raw_path.strip_prefix("data.").unwrap_or(raw_path);
        if path.is_empty() {
            self.data = value;
            return;
        }
        if !self.data.is_object() {
            self.data = serde_json::Value::Object(serde_json::Map::new());
        }
        let segs: Vec<&str> = path.split('.').collect();
        let mut cur = &mut self.data;
        for seg in &segs[..segs.len() - 1] {
            if !cur.is_object() {
                if let Some(map) = cur.as_array_mut() {
                    if let Ok(idx) = seg.parse::<usize>() {
                        if idx < map.len() {
                            cur = &mut map[idx];
                            continue;
                        }
                    }
                }
                return; // can't descend
            }
            let map = cur.as_object_mut().unwrap();
            if !map.contains_key(*seg) {
                map.insert(seg.to_string(), serde_json::json!({}));
            }
            cur = map.get_mut(*seg).unwrap();
        }
        let last = segs.last().unwrap();
        if let Some(map) = cur.as_object_mut() {
            map.insert(last.to_string(), value);
        } else if let Some(arr) = cur.as_array_mut() {
            if let Ok(idx) = last.parse::<usize>() {
                if idx < arr.len() {
                    arr[idx] = value;
                }
            }
        }
    }

    pub fn from_defaults(defaults: &serde_json::Value) -> Self {
        let mut store = Self::new();
        if let serde_json::Value::Object(map) = defaults {
            for (key, defn) in map {
                let default_val = if let serde_json::Value::Object(d) = defn {
                    d.get("default").cloned().unwrap_or(serde_json::Value::Null)
                } else {
                    defn.clone()
                };
                store.state.insert(key.clone(), default_val);
            }
        }
        store
    }

    // ── Dialog state ─────────────────────────────────────

    pub fn init_dialog(
        &mut self,
        dialog_id: &str,
        defaults: HashMap<String, serde_json::Value>,
        params: Option<HashMap<String, serde_json::Value>>,
    ) {
        self.dialog_id = Some(dialog_id.to_string());
        self.dialog = Some(defaults);
        self.dialog_params = params;
    }

    pub fn get_dialog(&self, key: &str) -> &serde_json::Value {
        self.dialog.as_ref()
            .and_then(|d| d.get(key))
            .unwrap_or(&serde_json::Value::Null)
    }

    pub fn set_dialog(&mut self, key: &str, value: serde_json::Value) {
        if let Some(ref mut d) = self.dialog {
            d.insert(key.to_string(), value);
            self.dialog_dirty = true;
        }
    }

    /// Set the dialog's on_change action name. Called by the
    /// open_dialog effect when the dialog spec declares an
    /// on_change field.
    pub fn set_dialog_on_change(&mut self, action: Option<String>) {
        self.dialog_on_change = action;
    }

    pub fn dialog_on_change(&self) -> Option<&str> {
        self.dialog_on_change.as_deref()
    }

    /// Take the dirty flag, leaving it false. Used by the post-run
    /// hook in run_effects to decide whether to fire on_change.
    pub fn take_dialog_dirty(&mut self) -> bool {
        let was = self.dialog_dirty;
        self.dialog_dirty = false;
        was
    }

    pub fn is_firing_on_change(&self) -> bool {
        self.dialog_firing_on_change
    }

    pub fn set_firing_on_change(&mut self, firing: bool) {
        self.dialog_firing_on_change = firing;
    }

    pub fn dialog_id(&self) -> Option<&str> {
        self.dialog_id.as_deref()
    }

    pub fn dialog_params(&self) -> Option<&HashMap<String, serde_json::Value>> {
        self.dialog_params.as_ref()
    }

    pub fn close_dialog(&mut self) {
        self.dialog_id = None;
        self.dialog = None;
        self.dialog_params = None;
        self.dialog_on_change = None;
        self.dialog_dirty = false;
        // dialog_firing_on_change is intentionally NOT cleared here —
        // close_dialog can be invoked from inside an on_change-fired
        // chain, and the guard must remain set until run_effects
        // unwinds.
    }

    /// Capture the current value of every state key referenced by a
    /// dialog's `preview_targets`. Phase 0 supports only top-level state
    /// keys (no dots in the path); deep paths are silently skipped and
    /// will land alongside their first real consumer in Phase 8/9.
    /// `targets` maps `dialog_state_key` → `state_key`.
    pub fn capture_dialog_snapshot(&mut self, targets: &HashMap<String, String>) {
        let mut snap = HashMap::new();
        for state_key in targets.values() {
            if !state_key.contains('.') {
                snap.insert(state_key.clone(), self.get(state_key).clone());
            }
        }
        self.dialog_snapshot = Some(snap);
    }

    pub fn dialog_snapshot(&self) -> Option<&HashMap<String, serde_json::Value>> {
        self.dialog_snapshot.as_ref()
    }

    pub fn clear_dialog_snapshot(&mut self) {
        self.dialog_snapshot = None;
    }

    pub fn has_dialog_snapshot(&self) -> bool {
        self.dialog_snapshot.is_some()
    }

    // ── Global state ─────────────────────────────────────

    pub fn get(&self, key: &str) -> &serde_json::Value {
        self.state.get(key).unwrap_or(&serde_json::Value::Null)
    }

    pub fn set(&mut self, key: &str, value: serde_json::Value) {
        self.state.insert(key.to_string(), value);
    }

    pub fn get_all(&self) -> &HashMap<String, serde_json::Value> {
        &self.state
    }

    // ── Panel state ──────────────────────────────────────

    pub fn init_panel(&mut self, panel_id: &str, defaults: HashMap<String, serde_json::Value>) {
        self.panels.insert(panel_id.to_string(), defaults);
    }

    pub fn get_panel(&self, panel_id: &str, key: &str) -> &serde_json::Value {
        self.panels.get(panel_id)
            .and_then(|p| p.get(key))
            .unwrap_or(&serde_json::Value::Null)
    }

    pub fn set_panel(&mut self, panel_id: &str, key: &str, value: serde_json::Value) {
        if let Some(scope) = self.panels.get_mut(panel_id) {
            scope.insert(key.to_string(), value);
        }
    }

    pub fn set_active_panel(&mut self, panel_id: Option<&str>) {
        self.active_panel = panel_id.map(|s| s.to_string());
    }

    pub fn active_panel_id(&self) -> Option<&str> {
        self.active_panel.as_deref()
    }

    pub fn destroy_panel(&mut self, panel_id: &str) {
        self.panels.remove(panel_id);
        if self.active_panel.as_deref() == Some(panel_id) {
            self.active_panel = None;
        }
    }

    // ── Tool state ───────────────────────────────────────
    //
    // Parallels `panel` scope but keyed by tool id (e.g. "selection").
    // YAML tool handlers write to `$tool.<id>.<key>`; the `set` effect
    // routes those targets here via [`set_tool`], and `eval_context`
    // exposes them under the `tool` key so expressions like
    // `$tool.selection.mode == "marquee"` resolve.

    /// Seed a tool's state namespace with `defaults` — typically the
    /// initial values declared in the tool's YAML `state:` block. Call
    /// once when a tool is registered.
    pub fn init_tool(&mut self, tool_id: &str, defaults: HashMap<String, serde_json::Value>) {
        self.tools.insert(tool_id.to_string(), defaults);
    }

    pub fn get_tool(&self, tool_id: &str, key: &str) -> &serde_json::Value {
        self.tools
            .get(tool_id)
            .and_then(|t| t.get(key))
            .unwrap_or(&serde_json::Value::Null)
    }

    pub fn set_tool(&mut self, tool_id: &str, key: &str, value: serde_json::Value) {
        // Create the tool namespace on first write — less friction for
        // callers that haven't explicitly run init_tool. Mirrors Flask
        // which auto-creates intermediates on dotted-path set().
        self.tools
            .entry(tool_id.to_string())
            .or_default()
            .insert(key.to_string(), value);
    }

    pub fn destroy_tool(&mut self, tool_id: &str) {
        self.tools.remove(tool_id);
    }

    /// Return the whole tool scope (for tests / inspection). The returned
    /// map is keyed by tool id; inner maps are key → value.
    pub fn tool_scopes(&self) -> &HashMap<String, HashMap<String, serde_json::Value>> {
        &self.tools
    }

    // ── List operations ──────────────────────────────────

    pub fn list_push(
        &mut self, panel_id: &str, key: &str, value: serde_json::Value,
        unique: bool, max_length: Option<usize>,
    ) {
        let scope = match self.panels.get_mut(panel_id) {
            Some(s) => s,
            None => return,
        };
        let lst = scope.entry(key.to_string())
            .or_insert_with(|| serde_json::Value::Array(vec![]));
        if let serde_json::Value::Array(arr) = lst {
            if unique {
                arr.retain(|item| item != &value);
            }
            arr.insert(0, value);
            if let Some(max) = max_length {
                arr.truncate(max);
            }
        }
    }

    // ── Context for expression evaluation ────────────────

    /// Build a serde_json::Value context for the expression evaluator.
    /// When a dialog is open, includes "dialog" and "param" keys.
    pub fn eval_context(&self) -> serde_json::Value {
        let mut ctx = serde_json::Map::new();

        // Global state
        let state_obj: serde_json::Value = serde_json::Value::Object(
            self.state.iter()
                .map(|(k, v)| (k.clone(), v.clone()))
                .collect()
        );
        ctx.insert("state".to_string(), state_obj);

        // Active panel state
        let panel_state = if let Some(pid) = &self.active_panel {
            if let Some(scope) = self.panels.get(pid) {
                serde_json::Value::Object(
                    scope.iter()
                        .map(|(k, v)| (k.clone(), v.clone()))
                        .collect()
                )
            } else {
                serde_json::Value::Object(serde_json::Map::new())
            }
        } else {
            serde_json::Value::Object(serde_json::Map::new())
        };
        ctx.insert("panel".to_string(), panel_state);

        // Tool state — one nested object per registered tool. Expressions
        // read this as `tool.<id>.<key>`.
        let tool_obj: serde_json::Map<String, serde_json::Value> = self
            .tools
            .iter()
            .map(|(tid, scope)| {
                let inner: serde_json::Map<String, serde_json::Value> = scope
                    .iter()
                    .map(|(k, v)| (k.clone(), v.clone()))
                    .collect();
                (tid.clone(), serde_json::Value::Object(inner))
            })
            .collect();
        ctx.insert("tool".to_string(), serde_json::Value::Object(tool_obj));

        // Dialog state
        if let Some(ref dialog) = self.dialog {
            let dialog_obj: serde_json::Value = serde_json::Value::Object(
                dialog.iter()
                    .map(|(k, v)| (k.clone(), v.clone()))
                    .collect()
            );
            ctx.insert("dialog".to_string(), dialog_obj);
        }

        // Dialog params
        if let Some(ref params) = self.dialog_params {
            let params_obj: serde_json::Value = serde_json::Value::Object(
                params.iter()
                    .map(|(k, v)| (k.clone(), v.clone()))
                    .collect()
            );
            ctx.insert("param".to_string(), params_obj);
        }

        serde_json::Value::Object(ctx)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn get_set() {
        let mut store = StateStore::new();
        store.set("x", serde_json::json!(5));
        assert_eq!(store.get("x"), &serde_json::json!(5));
    }

    #[test]
    fn missing_returns_null() {
        let store = StateStore::new();
        assert_eq!(store.get("missing"), &serde_json::Value::Null);
    }

    #[test]
    fn panel_scoping() {
        let mut store = StateStore::new();
        let mut c = HashMap::new();
        c.insert("mode".to_string(), serde_json::json!("hsb"));
        store.init_panel("color", c);

        let mut s = HashMap::new();
        s.insert("mode".to_string(), serde_json::json!("grid"));
        store.init_panel("swatches", s);

        assert_eq!(store.get_panel("color", "mode"), &serde_json::json!("hsb"));
        assert_eq!(store.get_panel("swatches", "mode"), &serde_json::json!("grid"));
    }

    #[test]
    fn eval_context_includes_panel() {
        let mut store = StateStore::new();
        store.set("fill_color", serde_json::json!("#ff0000"));
        let mut c = HashMap::new();
        c.insert("mode".to_string(), serde_json::json!("hsb"));
        store.init_panel("color", c);
        store.set_active_panel(Some("color"));

        let ctx = store.eval_context();
        assert_eq!(ctx["state"]["fill_color"], serde_json::json!("#ff0000"));
        assert_eq!(ctx["panel"]["mode"], serde_json::json!("hsb"));
    }

    #[test]
    fn list_push_unique() {
        let mut store = StateStore::new();
        let mut c = HashMap::new();
        c.insert("recent".to_string(), serde_json::json!(["a", "b", "c"]));
        store.init_panel("color", c);

        store.list_push("color", "recent", serde_json::json!("b"), true, Some(3));
        let result = store.get_panel("color", "recent");
        assert_eq!(result, &serde_json::json!(["b", "a", "c"]));
    }

    #[test]
    fn destroy_panel() {
        let mut store = StateStore::new();
        let mut c = HashMap::new();
        c.insert("x".to_string(), serde_json::json!(1));
        store.init_panel("test", c);
        store.destroy_panel("test");
        assert_eq!(store.get_panel("test", "x"), &serde_json::Value::Null);
    }

    #[test]
    fn dialog_init_and_get() {
        let mut store = StateStore::new();
        let mut defaults = HashMap::new();
        defaults.insert("h".to_string(), serde_json::json!(0));
        defaults.insert("color".to_string(), serde_json::json!("#ffffff"));
        let mut params = HashMap::new();
        params.insert("target".to_string(), serde_json::json!("fill"));
        store.init_dialog("color_picker", defaults, Some(params));
        assert_eq!(store.dialog_id(), Some("color_picker"));
        assert_eq!(store.get_dialog("h"), &serde_json::json!(0));
        assert_eq!(store.get_dialog("color"), &serde_json::json!("#ffffff"));
        assert_eq!(
            store.dialog_params().unwrap().get("target"),
            Some(&serde_json::json!("fill"))
        );
    }

    #[test]
    fn dialog_set() {
        let mut store = StateStore::new();
        let mut defaults = HashMap::new();
        defaults.insert("name".to_string(), serde_json::json!(""));
        store.init_dialog("test", defaults, None);
        store.set_dialog("name", serde_json::json!("hello"));
        assert_eq!(store.get_dialog("name"), &serde_json::json!("hello"));
    }

    #[test]
    fn dialog_close() {
        let mut store = StateStore::new();
        let mut defaults = HashMap::new();
        defaults.insert("x".to_string(), serde_json::json!(1));
        store.init_dialog("test", defaults, None);
        store.close_dialog();
        assert_eq!(store.dialog_id(), None);
        assert_eq!(store.get_dialog("x"), &serde_json::Value::Null);
    }

    #[test]
    fn eval_context_with_dialog() {
        let mut store = StateStore::new();
        store.set("fill_color", serde_json::json!("#ff0000"));
        let mut defaults = HashMap::new();
        defaults.insert("h".to_string(), serde_json::json!(180));
        let mut params = HashMap::new();
        params.insert("target".to_string(), serde_json::json!("fill"));
        store.init_dialog("picker", defaults, Some(params));
        let ctx = store.eval_context();
        assert_eq!(ctx["state"]["fill_color"], serde_json::json!("#ff0000"));
        assert_eq!(ctx["dialog"]["h"], serde_json::json!(180));
        assert_eq!(ctx["param"]["target"], serde_json::json!("fill"));
    }

    #[test]
    fn eval_context_no_dialog() {
        let store = StateStore::new();
        let ctx = store.eval_context();
        assert!(ctx.get("dialog").is_none());
        assert!(ctx.get("param").is_none());
    }

    #[test]
    fn dialog_and_panel_coexist() {
        let mut store = StateStore::new();
        let mut panel = HashMap::new();
        panel.insert("mode".to_string(), serde_json::json!("hsb"));
        store.init_panel("color", panel);
        store.set_active_panel(Some("color"));
        let mut dialog = HashMap::new();
        dialog.insert("h".to_string(), serde_json::json!(270));
        store.init_dialog("picker", dialog, None);
        let ctx = store.eval_context();
        assert_eq!(ctx["panel"]["mode"], serde_json::json!("hsb"));
        assert_eq!(ctx["dialog"]["h"], serde_json::json!(270));
    }

    // ── Preview snapshot/restore (Phase 0) ─────────────────────────
    //
    // capture_dialog_snapshot copies the current value of every state key
    // referenced by a dialog's preview_targets. Phase 0 supports only
    // top-level state keys (no path traversal); deep paths are skipped
    // and will be added in Phase 8/9 alongside their first real consumer.

    #[test]
    fn dialog_snapshot_capture_and_get() {
        let mut store = StateStore::new();
        store.set("left_indent", serde_json::json!(12));
        store.set("right_indent", serde_json::json!(0));
        let mut targets = HashMap::new();
        targets.insert("dlg_left".to_string(), "left_indent".to_string());
        targets.insert("dlg_right".to_string(), "right_indent".to_string());
        store.capture_dialog_snapshot(&targets);
        let snap = store.dialog_snapshot().expect("snapshot should be present");
        assert_eq!(snap.get("left_indent"), Some(&serde_json::json!(12)));
        assert_eq!(snap.get("right_indent"), Some(&serde_json::json!(0)));
        assert!(store.has_dialog_snapshot());
    }

    #[test]
    fn dialog_snapshot_clear_drops_it() {
        let mut store = StateStore::new();
        store.set("x", serde_json::json!(1));
        let mut targets = HashMap::new();
        targets.insert("k".to_string(), "x".to_string());
        store.capture_dialog_snapshot(&targets);
        assert!(store.has_dialog_snapshot());
        store.clear_dialog_snapshot();
        assert!(!store.has_dialog_snapshot());
        assert!(store.dialog_snapshot().is_none());
    }

    #[test]
    fn dialog_snapshot_skips_deep_paths_for_phase0() {
        let mut store = StateStore::new();
        store.set("flat", serde_json::json!(1));
        let mut targets = HashMap::new();
        targets.insert("a".to_string(), "flat".to_string());
        targets.insert("b".to_string(), "selection.deep.path".to_string());
        store.capture_dialog_snapshot(&targets);
        let snap = store.dialog_snapshot().expect("snapshot should be present");
        assert!(snap.contains_key("flat"));
        assert!(!snap.contains_key("selection.deep.path"));
    }
}
